compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp

```diff
@@ -124,85 +60,159 @@ auto unpackRegisteredMatchCallback(ImplicitLocOpBuilder &b,
 
- private:
-  void computeStrategy();
+  bool hasStage2() { return reductionTileSizeStage2.has_value(); }
+  int64_t getWarpShuffleSize() { return reductionTileSizeStage2.value(); }
+  int64_t getVectorSizeStage2() { return vectorSizeStage2.value(); }
 
-  int64_t reductionDimensionSize;
-  // TODO: Characterize shared memory consumption of this strategy and limit
-  // accordingly for good occupancy.
-  int64_t maxNumThreadsToUse;
+ private:
+  /// Maximal vector size (among {1, 2, 4}) that divides the
+  /// `reductionDimensionSize` and is used for vector transfers in Stage 1.
+  int64_t vectorSizeStage1;
+  /// Maximal "k-warp" size within the limits of the `maxNumThreadsToUse` and
+  /// `reductionDimensionSize` parameters.
+  /// This is also the blockDim.x of the kernel.
+  int64_t reductionTileSizeStage1;
+  /// Maximal vector size allowing to reduce from "k-warp" to single warp: when
+  /// `k` is a multiple of 4, choose 4; otherwise try with 2; otherwise just
+  /// use 1.
+  std::optional<int64_t> vectorSizeStage2;
+  /// `reductionTileSizeStage2` is exactly set to `warpShuffleSize` passed to
+  /// the constructor, only when reductionTileSizeStage1 > warpShuffleSize (i.e.
+  /// k > 1).
+  std::optional<int64_t> reductionTileSizeStage2;
+
+  /// Compute the staged strategy based on the `reductionDimensionSize`, the
+  /// `maxNumThreadsToUse` and the `warpShuffleSize`.
+  /// The latter 2 numbers control the tradeoff between parallelism and shared
+  /// memory consumption.
+  // TODO: Characterize shared memory consumption and limit for good occupancy.
+  // TODO: Support various elemental types.
+  void compute(int64_t reductionDimensionSize, int64_t maxNumThreadsToUse,
+               int64_t warpShuffleSize);
 };
 
-void ReductionStrategyThreadDistributionSizes::computeStrategy() {
-  vectorTileSize = 1;
-  reductionTileSize = maxNumThreadsToUse;
-  if (reductionDimensionSize <= 0) return;
-
-  // TODO: refine even further based on mod 2 and mod 4 only + min
-  // canonicalizations.
-  int64_t warpVector4Size = 4 * iree_compiler::kCudaWarpSize;
-  int64_t warpVector2Size = 2 * iree_compiler::kCudaWarpSize;
-  if (reductionDimensionSize % warpVector4Size == 0) {
-    int64_t f1 = reductionDimensionSize / warpVector4Size;
-    int64_t f2 = maxNumThreadsToUse / warpVector4Size;
-    reductionTileSize = std::min(f1, f2) * iree_compiler::kCudaWarpSize;
-    vectorTileSize = 4;
-  } else if (reductionDimensionSize % warpVector2Size == 0) {
-    int64_t f1 = reductionDimensionSize / warpVector2Size;
-    int64_t f2 = maxNumThreadsToUse / warpVector2Size;
-    reductionTileSize = std::min(f1, f2) * iree_compiler::kCudaWarpSize;
-    vectorTileSize = 2;
+static int64_t maxMultipleOf(int64_t val, int64_t multiple) {
+  assert(val > 0 && "expected nonnegative val");
+  assert(multiple > 0 && "expected nonnegative multiple");
+  return (val / multiple) * multiple;
+}
+
+static int64_t nextMultipleOf(int64_t val, int64_t multiple) {
```

Comment:
Document these plz. Especially `maxMultipleOf` sounds a bit confusing because it confuses the larger multiple of `multiple` that is _less than_ `val`. It's more STL-style `lower_bound`  than `max`.

Reply:
addressed in a followup